File: infra/src/stacks/prod.ts
import * as gcp from "@pulumi/gcp"
import { getEnvironmentStackConfig } from "../config"

const config = getEnvironmentStackConfig()

/**
 * ProdStack creates Cloud Run service for the production environment.
 * The service is referenced by the Serverless NEG created in shared stack.
 */

// Cloud Run V2 Service for prod
const webProd = new gcp.cloudrunv2.Service(
  "web-prod",
  {
    project: config.projectId,
    location: "asia-northeast1",
    name: "koborin-ai-web-prod",
    ingress: "INGRESS_TRAFFIC_INTERNAL_LOAD_BALANCER",
    template: {
      executionEnvironment: "EXECUTION_ENVIRONMENT_GEN2",
      containers: [
        {
          image: config.imageUri,
          envs: [
            // NODE_ENV: server-side environment detection
            { name: "NODE_ENV", value: "production" },
            // NEXT_PUBLIC_ENV: client-side environment detection (legacy, kept for compatibility)
            { name: "NEXT_PUBLIC_ENV", value: "prod" },
          ],
        },
      ],
      scaling: {
        minInstanceCount: 0,
        maxInstanceCount: 10,
      },
    },
    traffics: [
      {
        type: "TRAFFIC_TARGET_ALLOCATION_TYPE_LATEST",
        percent: 100,
      },
    ],
  },
  {
    // Import existing Cloud Run service from CDKTF state
    import: `projects/${config.projectId}/locations/asia-northeast1/services/koborin-ai-web-prod`,
  }
)

// Grant public access to prod Cloud Run service
const _webProdInvoker = new gcp.cloudrunv2.ServiceIamMember(
  "web-prod-invoker",
  {
    project: config.projectId,
    location: "asia-northeast1",
    name: webProd.name,
    role: "roles/run.invoker",
    member: "allUsers",
  },
  {
    // Import existing IAM binding from CDKTF state
    import: `projects/${config.projectId}/locations/asia-northeast1/services/koborin-ai-web-prod/roles/run.invoker/allUsers`,
  }
)

// ========================================
// Exports
// ========================================

export const serviceId = webProd.id
export const serviceName = webProd.name
export const serviceUri = webProd.uri

File: infra/stacks/prod.go
package stacks

import (
	"github.com/pulumi/pulumi-gcp/sdk/v8/go/gcp/cloudrunv2"
	"github.com/pulumi/pulumi/sdk/v3/go/pulumi"
	"github.com/pulumi/pulumi/sdk/v3/go/pulumi/config"
)

// Prod creates Cloud Run service for the production environment.
// The service is referenced by the Serverless NEG created in shared stack.
func Prod(ctx *pulumi.Context) error {
	cfg := config.New(ctx, "")
	gcpCfg := config.New(ctx, "gcp")

	projectID := gcpCfg.Require("project")
	imageURI := cfg.Require("imageUri")

	// Cloud Run V2 Service for prod
	webProd, err := cloudrunv2.NewService(ctx, "web-prod", &cloudrunv2.ServiceArgs{
		Project:  pulumi.String(projectID),
		Location: pulumi.String("asia-northeast1"),
		Name:     pulumi.String("koborin-ai-web-prod"),
		Ingress:  pulumi.String("INGRESS_TRAFFIC_INTERNAL_LOAD_BALANCER"),
		Template: &cloudrunv2.ServiceTemplateArgs{
			ExecutionEnvironment: pulumi.String("EXECUTION_ENVIRONMENT_GEN2"),
			Containers: cloudrunv2.ServiceTemplateContainerArray{
				&cloudrunv2.ServiceTemplateContainerArgs{
					Image: pulumi.String(imageURI),
					Envs: cloudrunv2.ServiceTemplateContainerEnvArray{
						// NODE_ENV: server-side environment detection
						&cloudrunv2.ServiceTemplateContainerEnvArgs{
							Name:  pulumi.String("NODE_ENV"),
							Value: pulumi.String("production"),
						},
						// NEXT_PUBLIC_ENV: client-side environment detection (legacy, kept for compatibility)
						&cloudrunv2.ServiceTemplateContainerEnvArgs{
							Name:  pulumi.String("NEXT_PUBLIC_ENV"),
							Value: pulumi.String("prod"),
						},
					},
				},
			},
			Scaling: &cloudrunv2.ServiceTemplateScalingArgs{
				MinInstanceCount: pulumi.Int(0),
				MaxInstanceCount: pulumi.Int(10),
			},
		},
		Traffics: cloudrunv2.ServiceTrafficArray{
			&cloudrunv2.ServiceTrafficArgs{
				Type:    pulumi.String("TRAFFIC_TARGET_ALLOCATION_TYPE_LATEST"),
				Percent: pulumi.Int(100),
			},
		},
	})
	if err != nil {
		return err
	}

	// Grant public access to prod Cloud Run service
	_, err = cloudrunv2.NewServiceIamMember(ctx, "web-prod-invoker", &cloudrunv2.ServiceIamMemberArgs{
		Project:  pulumi.String(projectID),
		Location: pulumi.String("asia-northeast1"),
		Name:     webProd.Name,
		Role:     pulumi.String("roles/run.invoker"),
		Member:   pulumi.String("allUsers"),
	})
	if err != nil {
		return err
	}

	return nil
}

